### Clearing and anlalyzing Autos.csv
The aim of the project is to clear and analyze the file "Autos.csv". 
The file encompases the information about selling and buying used cars from **from eBay Kleinanzeigen**. The data set includes 50,000 data points. 
Below is the header of the rows of the dataset.

| Column 	| Description 	|
|:-	|:-	|
| ateCrawled 	| When this ad was first crawled.  All field-values are taken from this date. 	|
| name 	| Name of the car. 	|
| seller 	| Whether the seller is private or a dealer. 	|
| offerType 	| The type of listing. 	|
| price 	| The price on the ad to sell the car. 	|
| abtest 	| Whether the listing is included in an A/B test. 	|
| vehicleType 	| The vehicle Type. 	|
| yearOfRegistration 	| The year in which the car was first registered. 	|
| gearbox 	| The transmission type. 	|
| powerPS 	| The power of the car in PS. 	|
| model 	| The car model name. 	|
| kilometer 	| How many kilometers the car has driven. 	|
| monthOfRegistration 	| The month in which the car was first registered. 	|
| fuelType 	| What type of fuel the car uses. 	|
| brand 	| The brand of the car. 	|
| notRepairedDamage 	| If the car has a damage which is not yet repaired. 	|

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding = "Windows-1252")

In [2]:
# quick overview of the dataframe
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

### Inspection of the Autos Dataframe

In [3]:
#checking general information about the DataFrame
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
#Viewing the first 5 rows
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### Observations about the dataframe
1. Non-null cells.Five columns of the dateset are integrers, wheres the rest 15 are objects. The column `vehicleType` has 44,905 non-null cells, which means that more that 5,000 rows do not have the type and may make the analysis more complicated. Columns `gearbox, model`, `fuelType`, `notRepairedDamage`. 
2. The header contains camelcase, whereas it would be better to write in a snakecase, instead of `yearOfRegistration: year_of_regisreation` and etc.
3. Many columns are in a text format: the `price` column has $ and `odometer` has km behind the number.

In [5]:
# let us see the content of coumns names
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [6]:
# we will define a function, which will replace the camelcase and will add an _ where there is no space. However, the easier solution is to create a new list, for example autos_col = [] and then to paste new names in there
def clean_columns(col):
  
    col = col.replace("yearOfRegistration", "registration_year")
    col = col.replace("monthOfRegistration", "registration_month")
    col = col.replace("notRepairedDamage", "unrepaired_damage")
    col = col.replace("dateCreated", "ad_created")
    col = col.replace("dateCrawled", "date_crawled")
    col = col.replace("offerType", "offer_type")
    col = col.replace("vehicleType", "vehicle_type")
    col = col.replace("gearbox", "gear_box")
    col = col.replace("nrOfPictures", "nr_pictures")
    col = col.replace("lastSeen", "last_seen")
    col = col.replace("postalCode", "postal_code")
    col = col.replace("powerPS", "power_ps")
    col = col.replace("fuelType", "fuel_type")
    return col

#creating a new list, in which we will update new colum names
new_columns = []
for c in autos.columns:
    update = clean_columns(c)
    new_columns.append(update)
#assigning the modified column names back     
autos.columns = new_columns 

#checking an updated dataframe
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

* The coumn labels were changed from camelcase to snakecase for the sake of better analysis in Pandas.
* We added underscore in between the names where there was no space to bring it to the standrardized view.


In [7]:
# Looking into descriptive statistics of all the colums
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In addition, we could have a look at the nr_pictures column to check if all the values are null.

In [8]:
autos["nr_pictures"].head(100)

0     0
1     0
2     0
3     0
4     0
     ..
95    0
96    0
97    0
98    0
99    0
Name: nr_pictures, Length: 100, dtype: int64

* In `offer_type` almost all values are the same.
* In `seller` columns almost all values are the same.
* The column `nr_pictures` has 0 in the column and , therefore, will not have any use for the analysis. It can be dropped.
* The `price` column has some 0 values as well as the $ ahead of the value, which must be removed.
* The `vehicle_type` has 44905 of 50000 vlues, whereas the others are NaN.
* The `registration_year` has an std of 105. This column requires clearing and more investigation.
* In `gear_box` 47320 cells are filled, whreas the others may be 0 or NaN.
* The column `model` has 47242 cells completed, the rest are 0 or NaN.
* `odometer` is stored as a text and required a removal of km and bringin to the numerical value.
* `fuel_type` column has some NaN or 0 cells. It requires checking.


### Clearing the `price` and `odometer` columns

We will start from `price` and `odometer`. We will remove any non-numeric carechters and will save as numeric dtype

In [9]:
#we use a str.replace method to remove all non-numeric values.astype is used to assign a numeric value
#price
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
#odometer
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

#odometer rename
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

#checking changes
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

    gear_box  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Next we will check `odometer_km` and `price` columns to explore them and to remove odd values

In [10]:

# we will check unique values
print(autos["odometer_km"].unique().shape)

# we will vie the statistical data info (min/max/mean)
print(autos["odometer_km"].describe())

# we will check the value counts
autos["odometer_km"].value_counts()

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In the `odometer` column, the values are reletively in order. It is worth pointing out that the majority of the entries in the `odometer` column are higher than 150 000 km. 

In [11]:
autos["odometer_km"].sort_index(ascending = True).head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [12]:
# we will check the value counts
print(autos["price"].value_counts(ascending=False))

# we will vie the statistical data info (min/max/mean)
print(autos["price"].describe())

# we will check the value counts
print(autos["price"].value_counts().sort_index(ascending = False).head(20))

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


In [13]:
# we will check the value counts
autos["price"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

The`price` column has got outliers. If we look into the  values in the descending order, we can see that there is a big debiation between the very top values \\$99999999 to \\$350000. After that the deviation becomes sequential and the price pattern does not see high price jumps. Therefore, we will keep the price range from \\$1 till \\$350000. 

In [14]:
autos = autos[(autos["price"] > 0) & (autos["price"] < 351000)]
autos["price"].describe()


count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Analyzing the date columns
The fllowing colums are all indentified as string values in pandas: `ate_crawled`, `last_seen`, `and ad_created`. We will need to convert then from strings into numerical representation to make an analysis.  

In [15]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

To undwerstand te date range, we will extract the date values from each column.

In [16]:
#Calculating the `date_crawled` distribution
print(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

#Sorting values according to distribution
print(autos["date_crawled"].str[:10].value_counts(sort=True, dropna=False))

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64
2016-04-03    1875
2016-03-20    1840
2016-03-21    1815
2016-03-12    1793
2016-03-14    1775
2016-04-04    1772
2016-03-07    1749
2016-04-02    1723
2016-03-28    1693
2016-03-19  

By observing the percentages of distribution, we can say that most of the ads were crawled between 2016-03-05  and 2016-04-07, with less activity seen at the end of the period.

In [17]:
# `ad_created` distribution
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

#Sorting values according to distribution
print(autos["ad_created"].str[:10].value_counts(sort=True, dropna=False))

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64
2016-04-03    1887
2016-03-20    1843
2016-03-21    1825
2016-04-04    1790
2016-03-12    1785
              ... 
2016-02-16       1
2016-02-07       1
2016-01-29       1
2016-01-14       1
2016-01-07       1
Name: ad_created, Length: 76, dtype: int64


We can observe that most of the ads were created between 2016-03-12 and 2016-04-03. 

In [18]:
# `last_seen` distribution
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

#Sorting values according to distribution
print(autos["last_seen"].str[:10].value_counts(sort=True, dropna=False))

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64
2016-04-06    10772
2016-04-07     6408
2016-04-05     6059
2016-03-17     1364
2016-04-03     1224
2016-04-02     1210
2016-03-30     1203
2016-04-04     1189
2016-03-31     1155
2016-0

By analysing the column `last_seen`, we can observe that most peak date is 2016-04-06. It may indicate that most of the transactions  occured on that date or in the nearby dates.

### Analysing the `registration_year` 
We will have a look in details into this column

In [19]:
# checking the distribution of the `registration_year` column
autos['registration_year']. describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The `registration_year` distribution information shows that this column must be cleaned, as the mimimum value is 1000 and max is 9999.

Let us look into the distribution of the registration year so that we can understand what year range we should select.

In [20]:
# checking the earliest registration year
print(autos["registration_year"].value_counts().sort_index().head(30))

#Sorting values according to distribution
print(autos["registration_year"].value_counts().sort_index().tail(30))


1000     1
1001     1
1111     1
1800     2
1910     5
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     4
1957     2
1958     4
1959     6
1960    23
1961     6
1962     4
1963     8
1964    12
Name: registration_year, dtype: int64
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64


- According to the above chart, 1910 would be the most possible regisrration year. The years listed earlier will be dropped.
- We will select 2016 as the highest registration year, because the information was gathered by 2016 latest.

Let us drop the rest of the years, which would be considered as inacurate and would hinder the analysis.

In [21]:
#deleting the odd registration year range
autos = autos[(autos["registration_year"] >= 1910) & (autos["registration_year"] <= 2016)]

autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [22]:
# calculating the distribution of the remaining values of the `registration_year`
autos["registration_year"].value_counts(normalize=True).head(40)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
1980    0.001821
1987    0.001542
1986    0.001542
1983    0.001093
1984    0.001093
1978    0.000943
1982    0.000878
1970    0.000814
1979    0.000728
1972    0.000707
Name: registration_year, dtype: float64

As we can see, the majority of the cars were registered within the last 25 years, i.e. in the range from 1993 till 2016. 

In [23]:
autos["registration_year"].value_counts(normalize=True).head(25).sum()

0.9575201902272874

95.7% of all the cars have got the registration year between 1993 and 2016. The registration years, whose percenrage was 1 and higher were taken into account.

### Exploring the `brand` column

In [24]:
#we will display the description of the `brand` column
print(autos["brand"].describe())

# calculating distribution
autos_brand = autos["brand"].value_counts(normalize=True)
print(autos_brand)

#showing the 5% of the most sold brands
brand_best = autos_brand[autos_brand > 0.01].index
print(brand_best)

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.00209

The most sold brands are as follows:
1. Volkswagen     21.1%
2. BMW            11.0%
3. Opel           10.7%
4. Mercedes_benz  9.6%
5. Audi           8.6%
6. Ford           6.9%
More that 60% of all car sales are of German brand.     
The next "foreign" car in the group is "Ford".

In [25]:
selected_price = {}

for p in brand_best:
    top_brands_selected = autos[autos["brand"] == p]
    mean_prices = top_brands_selected["price"].mean().round(decimals=2)
    selected_price[p] = mean_prices
selected_price


{'volkswagen': 5402.41,
 'bmw': 8332.82,
 'opel': 2975.24,
 'mercedes_benz': 8628.45,
 'audi': 9336.69,
 'ford': 3749.47,
 'renault': 2474.86,
 'peugeot': 3094.02,
 'fiat': 2813.75,
 'seat': 4397.23,
 'skoda': 6368.0,
 'nissan': 4743.4,
 'mazda': 4112.6,
 'smart': 3580.22,
 'citroen': 3779.14,
 'toyota': 5167.09,
 'hyundai': 5365.25}

- __Highest price__. Audi and Mercedez and BMW are sold at the highet prices. Thus, the avargae price for Audi is \\$9336. BMW and Mercedec price offers re 2nd and 3rd acordingly. If we compare the rate of brands introduced in the brand range, Audi's share is 8.6 percent. This may indicate that the demand for this brand is relatively high, despite the fact that it is considered as higher-than-average car. 

- __The most suitable car__. VW's prices are on the average range \\$5402, with the brand offer range being at 21%. This may indicate that this brand meets the quantity/price needs of the buyers. 


### Relationship between milage and price
In the next section we will cnsider where there is a correspondance between the odometer column and prices.

In [26]:
# creating a dictionary for the top brands by milage
brand_milage = {}

for b in brand_best:
    top_brands_selected = autos[autos["brand"] == b]
    milage_selected = top_brands_selected["odometer_km"].mean().round(decimals=0)
    brand_milage[b] = milage_selected
brand_milage

{'volkswagen': 128707.0,
 'bmw': 132573.0,
 'opel': 129310.0,
 'mercedes_benz': 130788.0,
 'audi': 129157.0,
 'ford': 124266.0,
 'renault': 128071.0,
 'peugeot': 127154.0,
 'fiat': 117122.0,
 'seat': 121131.0,
 'skoda': 110849.0,
 'nissan': 118331.0,
 'mazda': 124464.0,
 'smart': 99327.0,
 'citroen': 119694.0,
 'toyota': 115944.0,
 'hyundai': 106442.0}

The next step is to convert both price and milage dictionaries.

In [27]:
# conversion of `selected_price`
selected_price_series = pd.Series(selected_price).sort_values(ascending=False)
print(selected_price_series)

# conversion of `selected_milage`
selected_brand_milage = pd.Series(brand_milage).sort_values(ascending=False)
print(selected_brand_milage)

audi             9336.69
mercedes_benz    8628.45
bmw              8332.82
skoda            6368.00
volkswagen       5402.41
hyundai          5365.25
toyota           5167.09
nissan           4743.40
seat             4397.23
mazda            4112.60
citroen          3779.14
ford             3749.47
smart            3580.22
peugeot          3094.02
opel             2975.24
fiat             2813.75
renault          2474.86
dtype: float64
bmw              132573.0
mercedes_benz    130788.0
opel             129310.0
audi             129157.0
volkswagen       128707.0
renault          128071.0
peugeot          127154.0
mazda            124464.0
ford             124266.0
seat             121131.0
citroen          119694.0
nissan           118331.0
fiat             117122.0
toyota           115944.0
skoda            110849.0
hyundai          106442.0
smart             99327.0
dtype: float64


In [29]:
#creating a new dataframe out of the series
df = pd.DataFrame(selected_price_series, columns=["mean_price"])

df["mean_milage"] = selected_brand_milage
df


mean_price  mean_milage
audi              9336.69     129157.0
mercedes_benz     8628.45     130788.0
bmw               8332.82     132573.0
skoda             6368.00     110849.0
volkswagen        5402.41     128707.0
hyundai           5365.25     106442.0
toyota            5167.09     115944.0
nissan            4743.40     118331.0
seat              4397.23     121131.0
mazda             4112.60     124464.0
citroen           3779.14     119694.0
ford              3749.47     124266.0
smart             3580.22      99327.0
peugeot           3094.02     127154.0
opel              2975.24     129310.0
fiat              2813.75     117122.0
renault           2474.86     128071.0

We can conclude that in this case no correlation between the ilage and price can be seen. Let's consider the following pattern. 

There the mean price for first theree brands __Audi__, __Mercedez_Benz__, __BMW__ remains high, despite the relatively high odometer values. This may bring to conclusion that the selection of the vehicle happens mostly on the name of the brand and, possibily, the registration year. 

###Conclusions

- 95.7% of all the cars have got the registration year between 1993 and 2016. 
- German brands remain the most popular, with buyers being concentrated on acuiring high premium brands: __Audi__, __Mercedes_Benz__, and __BMW__.
- The most sold brands are as follows:
1. Volkswagen     21.1%
2. BMW            11.0%
3. Opel           10.7%
4. Mercedes_benz  9.6%
5. Audi           8.6%
6. Ford           6.9%
More that 60% of all car sales are of German brand.     
The next "foreign" car in the group is "Ford".
- __The most suitable car__. VW's prices are on the average range \\$5402, with the brand offer range being at 21%. This may indicate that this brand meets the quantity/price needs of the buyers.